consensus/abci.go

```diff
@@ -97,11 +98,24 @@ func (abci *TendermintABCI) InitChain(req tmtAbciTypes.RequestInitChain) tmtAbci
 // for the validators.
 //
 // Response:
-//		- Optional Key-Value tags for filtering and indexing
+// 		- Optional Key-Value tags for filtering and indexing
 func (abci *TendermintABCI) BeginBlock(req tmtAbciTypes.RequestBeginBlock) tmtAbciTypes.ResponseBeginBlock {
-	abci.logger.Debug("Beginning new block", "hash", req.Hash)
-	abci.db.UpdateBlockState(&req.Header)
-
+	abci.logger.Debug("Beginning new block", "hash", req.Hash, "height", req.Header.Height)
+	parentBlock := abci.getCurrentBlock()
+
+	// IMPORTANT: According to Tendermint documentation and based on the consensus setup made for our network
+	// two consecutive blocks cannot generated less than one second apart. BUT we identify an issue on that assumption
+	// and reported here https://github.com/tendermint/tendermint/issues/3755. Therefore we implemented this mitigation code
+	// to prevent ethereum headers to be invalid. Learn more in https://github.com/lightstreams-network/lightchain/issues/186
+	if uint64(req.Header.Time.Unix()) <= parentBlock.Time() {
+		abci.metrics.ReplacedBlockTimeTotal.Add(1)
+		nextBlockTime := time.Unix(int64(parentBlock.Time() + 1), 0)
+		abci.logger.Error(fmt.Sprintf("Invalid consensus BlockTime. Replacing block time %d...", req.Header.Height), 
+			"original", req.Header.Time.Unix(), "replaced", nextBlockTime.Unix())
+		req.Header.Time = nextBlockTime
+	}
+	
+	abci.db.UpdateBlockState(req.Header)
```

Comment:
I improved here the logging a bit to be more consistent and to avoid printing raw bytes around: https://github.com/lightstreams-network/lightchain/pull/187/commits/df8ca57518db776e74705e3ced252505b8476324